sdk/eventgrid/azure-eventgrid/azure/eventgrid/_helpers.py

```diff
@@ -0,0 +1,72 @@
+import hashlib
+import hmac
+import base64
+try:
+    from urllib.parse import quote
+except ImportError:
+    from urllib2 import quote # type: ignore
+import datetime
+
+from azure.core.pipeline.policies import AzureKeyCredentialPolicy
+from azure.core.credentials import AzureKeyCredential
+from ._shared_access_signature_credential import EventGridSharedAccessSignatureCredential
+from ._signature_credential_policy import EventGridSharedAccessSignatureCredentialPolicy
+from . import _constants as constants
+
+def generate_shared_access_signature(topic_hostname, shared_access_key, expiration_date_utc, **kwargs):
+    # type: (str, str, datetime.Datetime, Any) -> str
+    """ Helper method to generate shared access signature given hostname, key, and expiration date.
+        :param str topic_hostname: The topic endpoint to send the events to.
+            Similar to <YOUR-TOPIC-NAME>.<YOUR-REGION-NAME>-1.eventgrid.azure.net
+        :param str shared_access_key: The shared access key to be used for generating the token
+        :param datetime.datetime expiration_date_utc: The expiration datetime in UTC for the signature.
+        :param str api_version: The API Version to include in the signature. If not provided, the default API version will be used.
+        :rtype: str
+    """
+
+    full_topic_hostname = _get_full_topic_hostname(topic_hostname)
+
+    full_topic_hostname = "{}?apiVersion={}".format(full_topic_hostname, kwargs.get('api_version', None) or constants.DEFAULT_API_VERSION)
+    encoded_resource = quote(full_topic_hostname, safe=constants.SAFE_ENCODE)
+    encoded_expiration_utc = quote(str(expiration_date_utc), safe=constants.SAFE_ENCODE)
+
+    unsigned_sas = "r={}&e={}".format(encoded_resource, encoded_expiration_utc)
+    signature = quote(_hmac(shared_access_key, unsigned_sas), safe=constants.SAFE_ENCODE)
+    signed_sas = "{}&s={}".format(unsigned_sas, signature)
+    return signed_sas
+
+def _get_topic_hostname_only_fqdn(topic_hostname):
+    if topic_hostname.startswith('http://'):
+        raise ValueError("HTTP is not supported. Only HTTPS is supported.")
+    if topic_hostname.startswith('https://'):
+        topic_hostname = topic_hostname.replace("https://", "")
+    if topic_hostname.endswith("/api/events"):
+        topic_hostname = topic_hostname.replace("/api/events", "")
+    
+    return topic_hostname
+
+def _get_full_topic_hostname(topic_hostname):
+    if topic_hostname.startswith('http://'):
+        raise ValueError("HTTP is not supported. Only HTTPS is supported.")
+    if not topic_hostname.startswith('https://'):
+        topic_hostname = "https://{}".format(topic_hostname)
+    if not topic_hostname.endswith("/api/events"):
+        topic_hostname = "{}/api/events".format(topic_hostname)
+    
+    return topic_hostname
+
+def _hmac(key, message):
```

Comment:
nit: functions should have imperitive names; so "generate_hmac"?

Reply:
quick 2nd note on this:  There's a push to have type signatures for internal functions.  I'm not gonna fine-toothed-comb on this now since we're in a very early preview, but my learnings from SB is that life is a lot easier of one just does the typing/docstring bits as one goes along.